In [1]:
documents = [
    "This is a list which containing sample documents.",
    "Keywords are important for keyword-based search.",
    "Document analysis involves extracting keywords.",
    "Keyword-based search relies on sparse embeddings.",
    "Understanding document structure aids in keyword extraction.",
    "Efficient keyword extraction enhances search accuracy.",
    "Semantic similarity improves document retrieval performance.",
    "Machine learning algorithms can optimize keyword extraction methods."
]

In [ ]:
!pip install sentence_transformers

In [2]:
from sentence_transformers import SentenceTransformer

In [3]:
# Load pre-trained Sentence Transformer model
model_name = 'sentence-transformers/paraphrase-xlm-r-multilingual-v1'

In [4]:
model = SentenceTransformer(model_name)

In [5]:
documents

['This is a list which containing sample documents.',
 'Keywords are important for keyword-based search.',
 'Document analysis involves extracting keywords.',
 'Keyword-based search relies on sparse embeddings.',
 'Understanding document structure aids in keyword extraction.',
 'Efficient keyword extraction enhances search accuracy.',
 'Semantic similarity improves document retrieval performance.',
 'Machine learning algorithms can optimize keyword extraction methods.']

In [6]:
len(documents)

8

In [7]:
document_embeddings = model.encode(documents)

In [8]:
len(document_embeddings)

8

In [9]:
len(document_embeddings[0])

768

In [10]:
for i, embedding in enumerate(document_embeddings):
    print(f"Document {i+1} embedding: {embedding}")

Document 1 embedding: [ 0.10894685  0.0781207   0.11626569 -0.31912503  0.4689022   0.43514383
  0.01453738  0.44238767  0.297164   -0.18982707  0.07389054 -0.27864906
  0.21338163 -0.1207701   0.17891699 -0.00789878  0.04754862 -0.18204565
  0.3422711  -0.06994258 -0.14288741  0.5714125  -0.11153244 -0.17895408
  0.01523129  0.2610571  -0.20555831  0.05203114 -0.02810769  0.2387324
  0.0120698   0.04404927  0.02242311 -0.13895182 -0.7410038   0.25601006
  0.08149689  0.18820496 -0.4123769   0.11368618  0.28121173  0.05860886
 -0.1731878   0.33549133  0.21803683 -0.05090715 -0.05457789 -0.8738479
 -0.24082275  0.32006976  0.44761655  0.06347825  0.5357486   0.16607258
 -0.33197     0.33393636  0.2861592  -0.5419566  -0.2713242   0.24881154
 -0.23919384 -0.4692628   0.13836576  0.3784289  -0.01304429  0.019906
  0.32365087  0.45857537  0.07600261  0.25299588 -0.42938945  0.10051926
 -0.33042598 -0.6987648   0.01035903  0.05666573  0.14731245 -0.47082353
  0.08063988  0.33870456 -0.27278

In [12]:
from pprint import pprint

pprint(document_embeddings.tolist())

[[0.10894685238599777,
  0.07812070101499557,
  0.1162656918168068,
  -0.3191250264644623,
  0.46890220046043396,
  0.4351438283920288,
  0.014537381939589977,
  0.4423876702785492,
  0.2971639931201935,
  -0.1898270696401596,
  0.07389053702354431,
  -0.27864906191825867,
  0.21338163316249847,
  -0.12077010422945023,
  0.17891699075698853,
  -0.007898781448602676,
  0.04754862189292908,
  -0.18204565346240997,
  0.342271089553833,
  -0.06994257867336273,
  -0.1428874135017395,
  0.5714125037193298,
  -0.11153244227170944,
  -0.1789540797472,
  0.015231293626129627,
  0.2610571086406708,
  -0.20555831491947174,
  0.05203114077448845,
  -0.028107689693570137,
  0.23873239755630493,
  0.012069798074662685,
  0.04404927417635918,
  0.02242310531437397,
  -0.13895182311534882,
  -0.7410038113594055,
  0.2560100555419922,
  0.08149689435958862,
  0.18820495903491974,
  -0.41237691044807434,
  0.11368618160486221,
  0.2812117338180542,
  0.05860886350274086,
  -0.17318780720233917,
  0.3354

In [13]:
from pprint import pprint

pprint(document_embeddings[0].tolist())

[0.10894685238599777,
 0.07812070101499557,
 0.1162656918168068,
 -0.3191250264644623,
 0.46890220046043396,
 0.4351438283920288,
 0.014537381939589977,
 0.4423876702785492,
 0.2971639931201935,
 -0.1898270696401596,
 0.07389053702354431,
 -0.27864906191825867,
 0.21338163316249847,
 -0.12077010422945023,
 0.17891699075698853,
 -0.007898781448602676,
 0.04754862189292908,
 -0.18204565346240997,
 0.342271089553833,
 -0.06994257867336273,
 -0.1428874135017395,
 0.5714125037193298,
 -0.11153244227170944,
 -0.1789540797472,
 0.015231293626129627,
 0.2610571086406708,
 -0.20555831491947174,
 0.05203114077448845,
 -0.028107689693570137,
 0.23873239755630493,
 0.012069798074662685,
 0.04404927417635918,
 0.02242310531437397,
 -0.13895182311534882,
 -0.7410038113594055,
 0.2560100555419922,
 0.08149689435958862,
 0.18820495903491974,
 -0.41237691044807434,
 0.11368618160486221,
 0.2812117338180542,
 0.05860886350274086,
 -0.17318780720233917,
 0.3354913294315338,
 0.21803683042526245,
 -0.0509

In [14]:
documents

['This is a list which containing sample documents.',
 'Keywords are important for keyword-based search.',
 'Document analysis involves extracting keywords.',
 'Keyword-based search relies on sparse embeddings.',
 'Understanding document structure aids in keyword extraction.',
 'Efficient keyword extraction enhances search accuracy.',
 'Semantic similarity improves document retrieval performance.',
 'Machine learning algorithms can optimize keyword extraction methods.']

In [15]:
query = "Natural language processing techniques enhance keyword extraction efficiency."

In [16]:
query_embedding = model.encode(query)

In [17]:
print("Query embedding:", query_embedding)

Query embedding: [ 3.42821002e-01  3.26170415e-01  7.73259671e-03  1.60758961e-02
  2.38518342e-02 -2.52880692e-01  8.15468654e-02  1.37177244e-01
  3.33291739e-01  3.65743861e-02 -9.28225219e-02 -3.43725860e-01
  7.12015778e-02  9.48833764e-01 -9.41289589e-02 -7.96414018e-01
 -3.20592105e-01 -2.67564505e-01  2.08664313e-01  7.99598172e-02
  7.93332979e-03 -5.22297025e-02 -1.00512810e-01 -2.04229861e-01
  2.19562590e-01  7.67739192e-02  1.25881582e-01  2.90843904e-01
 -4.99191016e-01  2.27709681e-01 -1.94742501e-01  7.21394569e-02
 -4.56504943e-03  1.21945590e-01 -1.46225840e-01 -9.64745134e-02
  1.02258325e-01 -2.04421028e-01 -3.95657808e-01  7.32252449e-02
  1.10104240e-01  4.84611094e-01  1.04519300e-01  3.86131763e-01
 -1.68697283e-01 -2.11251169e-01 -6.38036370e-01 -4.32641506e-01
  2.74459094e-01  9.64054242e-02 -1.30053580e-01  5.76566905e-02
  8.07962045e-02  6.48147911e-02 -2.14187102e-03  2.29773849e-01
 -6.66053444e-02  3.07675540e-01  2.11779416e-01 -1.57844543e-01
  9.7885

In [18]:
len(query_embedding)

768

In [21]:
query_embedding.shape

(768,)

In [22]:
np.array([query_embedding]).shape

(1, 768)

In [19]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

In [20]:
similarities = cosine_similarity(np.array([query_embedding]), document_embeddings)

In [23]:
similarities

array([[0.16948149, 0.45802277, 0.56756943, 0.44123298, 0.63161176,
        0.75214124, 0.5503521 , 0.7448166 ]], dtype=float32)

In [24]:
print(similarities.shape)

(1, 8)


In [27]:
most_similar_index = np.argmax(similarities)
most_similar_index

np.int64(5)

In [28]:
most_similar_index = int(np.argmax(similarities))
most_similar_index

5

In [29]:
similarity_score = float(similarities[0][most_similar_index])
similarity_score

0.7521412372589111

In [30]:
most_similar_document = documents[most_similar_index]
most_similar_document

'Efficient keyword extraction enhances search accuracy.'

In [ ]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

'Natural language processing techniques enhance keyword extraction efficiency.'

In [33]:
similarity_score = similarities[0][most_similar_index]
similarity_score

np.float32(0.75214124)

In [34]:
sorted_indices = np.argsort(similarities[0])[::-1]
sorted_indices

array([5, 7, 4, 2, 6, 1, 3, 0])

In [35]:
ranked_documents = [(documents[i], similarities[0][i]) for i in sorted_indices]
ranked_documents

[('Efficient keyword extraction enhances search accuracy.',
  np.float32(0.75214124)),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  np.float32(0.7448166)),
 ('Understanding document structure aids in keyword extraction.',
  np.float32(0.63161176)),
 ('Document analysis involves extracting keywords.', np.float32(0.56756943)),
 ('Semantic similarity improves document retrieval performance.',
  np.float32(0.5503521)),
 ('Keywords are important for keyword-based search.', np.float32(0.45802277)),
 ('Keyword-based search relies on sparse embeddings.', np.float32(0.44123298)),
 ('This is a list which containing sample documents.', np.float32(0.16948149))]

In [36]:
ranked_documents = [(documents[i], similarities[0][i]) for i in sorted_indices]
ranked_documents

[('Efficient keyword extraction enhances search accuracy.',
  np.float32(0.75214124)),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  np.float32(0.7448166)),
 ('Understanding document structure aids in keyword extraction.',
  np.float32(0.63161176)),
 ('Document analysis involves extracting keywords.', np.float32(0.56756943)),
 ('Semantic similarity improves document retrieval performance.',
  np.float32(0.5503521)),
 ('Keywords are important for keyword-based search.', np.float32(0.45802277)),
 ('Keyword-based search relies on sparse embeddings.', np.float32(0.44123298)),
 ('This is a list which containing sample documents.', np.float32(0.16948149))]

In [37]:
ranked_documents = [
    (documents[i], float(similarities[0][i]))
    for i in sorted_indices
]

ranked_documents

[('Efficient keyword extraction enhances search accuracy.',
  0.7521412372589111),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  0.7448166012763977),
 ('Understanding document structure aids in keyword extraction.',
  0.6316117644309998),
 ('Document analysis involves extracting keywords.', 0.5675694346427917),
 ('Semantic similarity improves document retrieval performance.',
  0.5503520965576172),
 ('Keywords are important for keyword-based search.', 0.4580227732658386),
 ('Keyword-based search relies on sparse embeddings.', 0.44123297929763794),
 ('This is a list which containing sample documents.', 0.16948148608207703)]

In [38]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

In [39]:
print("Ranked Documents:")
for rank, (document, similarity) in enumerate(ranked_documents, start=1):
    print(f"Rank {rank}: Document - '{document}', Similarity Score - {similarity}")

Ranked Documents:
Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 0.7521412372589111
Rank 2: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.7448166012763977
Rank 3: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - 0.6316117644309998
Rank 4: Document - 'Document analysis involves extracting keywords.', Similarity Score - 0.5675694346427917
Rank 5: Document - 'Semantic similarity improves document retrieval performance.', Similarity Score - 0.5503520965576172
Rank 6: Document - 'Keywords are important for keyword-based search.', Similarity Score - 0.4580227732658386
Rank 7: Document - 'Keyword-based search relies on sparse embeddings.', Similarity Score - 0.44123297929763794
Rank 8: Document - 'This is a list which containing sample documents.', Similarity Score - 0.16948148608207703


In [40]:
print("Top 4 Documents:")
for rank, (document, similarity) in enumerate(ranked_documents[:4], start=1):
    print(f"Rank {rank}: Document - '{document}', Similarity Score - {similarity}")

Top 4 Documents:
Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 0.7521412372589111
Rank 2: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.7448166012763977
Rank 3: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - 0.6316117644309998
Rank 4: Document - 'Document analysis involves extracting keywords.', Similarity Score - 0.5675694346427917


# BM25 Reranking

In [41]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

In [ ]:
!pip install rank_bm25

In [42]:
from rank_bm25 import BM25Okapi

In [43]:
top_4_documents = [doc[0] for doc in ranked_documents[:4]]

In [44]:
top_4_documents

['Efficient keyword extraction enhances search accuracy.',
 'Machine learning algorithms can optimize keyword extraction methods.',
 'Understanding document structure aids in keyword extraction.',
 'Document analysis involves extracting keywords.']

In [45]:
tokenized_top_4_documents = [doc.split() for doc in top_4_documents]
tokenized_top_4_documents

[['Efficient', 'keyword', 'extraction', 'enhances', 'search', 'accuracy.'],
 ['Machine',
  'learning',
  'algorithms',
  'can',
  'optimize',
  'keyword',
  'extraction',
  'methods.'],
 ['Understanding',
  'document',
  'structure',
  'aids',
  'in',
  'keyword',
  'extraction.'],
 ['Document', 'analysis', 'involves', 'extracting', 'keywords.']]

In [46]:
tokenized_query = query.split()

In [47]:
tokenized_query

['Natural',
 'language',
 'processing',
 'techniques',
 'enhance',
 'keyword',
 'extraction',
 'efficiency.']

In [48]:
bm25=BM25Okapi(tokenized_top_4_documents)
bm25

In [49]:
bm25_scores = bm25.get_scores(tokenized_query)
bm25_scores

array([0.1907998 , 0.16686672, 0.17803252, 0.        ])

In [50]:
sorted_indices2 = np.argsort(bm25_scores)[::-1]
sorted_indices2

array([0, 2, 1, 3])

In [51]:
top_4_documents

['Efficient keyword extraction enhances search accuracy.',
 'Machine learning algorithms can optimize keyword extraction methods.',
 'Understanding document structure aids in keyword extraction.',
 'Document analysis involves extracting keywords.']

In [52]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

In [53]:
reranked_documents = [(top_4_documents[i], bm25_scores[i]) for i in sorted_indices2]
reranked_documents

[('Efficient keyword extraction enhances search accuracy.',
  np.float64(0.19079979534096053)),
 ('Understanding document structure aids in keyword extraction.',
  np.float64(0.1780325227902643)),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  np.float64(0.1668667199671815)),
 ('Document analysis involves extracting keywords.', np.float64(0.0))]

In [54]:
#Converting the score to float
reranked_documents = [
    (top_4_documents[i], float(bm25_scores[i]))
    for i in sorted_indices2
]
reranked_documents

[('Efficient keyword extraction enhances search accuracy.',
  0.19079979534096053),
 ('Understanding document structure aids in keyword extraction.',
  0.1780325227902643),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  0.1668667199671815),
 ('Document analysis involves extracting keywords.', 0.0)]

In [55]:
print("Rerank of top 4 Documents:")
for rank, (document, similarity) in enumerate(reranked_documents, start=1):
    print(f"Rank {rank}: Document - '{document}', Similarity Score - {similarity}")

Rerank of top 4 Documents:
Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 0.19079979534096053
Rank 2: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - 0.1780325227902643
Rank 3: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.1668667199671815
Rank 4: Document - 'Document analysis involves extracting keywords.', Similarity Score - 0.0


In [56]:
ranked_documents[:4]

[('Efficient keyword extraction enhances search accuracy.',
  0.7521412372589111),
 ('Machine learning algorithms can optimize keyword extraction methods.',
  0.7448166012763977),
 ('Understanding document structure aids in keyword extraction.',
  0.6316117644309998),
 ('Document analysis involves extracting keywords.', 0.5675694346427917)]

# Cross-Encoder Reranking

In [57]:
from sentence_transformers import CrossEncoder

In [58]:
cross_encoder = CrossEncoder('cross-encoder/ms-marco-MiniLM-L-6-v2')

config.json:   0%|          | 0.00/794 [00:00<?, ?B/s]

f:\Advance RAG\rag_env\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Lenovo\.cache\huggingface\hub\models--cross-encoder--ms-marco-MiniLM-L-6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTT

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

In [124]:
top_4_documents

['Efficient keyword extraction enhances search accuracy.',
 'Machine learning algorithms can optimize keyword extraction methods.',
 'Understanding document structure aids in keyword extraction.',
 'Document analysis involves extracting keywords.']

In [125]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

In [126]:
pairs = []
for doc in top_4_documents:
    pairs.append([query, doc])

In [127]:
pairs

[['Natural language processing techniques enhance keyword extraction efficiency.',
  'Efficient keyword extraction enhances search accuracy.'],
 ['Natural language processing techniques enhance keyword extraction efficiency.',
  'Machine learning algorithms can optimize keyword extraction methods.'],
 ['Natural language processing techniques enhance keyword extraction efficiency.',
  'Understanding document structure aids in keyword extraction.'],
 ['Natural language processing techniques enhance keyword extraction efficiency.',
  'Document analysis involves extracting keywords.']]

In [128]:
scores = cross_encoder.predict(pairs)
scores

array([ 3.1378732,  0.8421646, -2.9193015, -2.8781915], dtype=float32)

In [129]:
scored_docs = zip(scores, top_4_documents)

In [130]:
scored_docs

In [131]:
reranked_document_cross_encoder = sorted(scored_docs, reverse=True)
reranked_document_cross_encoder

[(np.float32(3.1378732),
  'Efficient keyword extraction enhances search accuracy.'),
 (np.float32(0.8421646),
  'Machine learning algorithms can optimize keyword extraction methods.'),
 (np.float32(-2.8781915), 'Document analysis involves extracting keywords.'),
 (np.float32(-2.9193015),
  'Understanding document structure aids in keyword extraction.')]

# Cohere API Reranking


In [ ]:
!pip install cohere

In [67]:
import cohere

In [69]:
from dotenv import load_dotenv
import os

# Load variables from .env
load_dotenv()

# Get the API key
cohere_api_key = os.getenv("cohere_api_key")


In [74]:
co = cohere.Client(cohere_api_key)

In [70]:
top_4_documents

['Efficient keyword extraction enhances search accuracy.',
 'Machine learning algorithms can optimize keyword extraction methods.',
 'Understanding document structure aids in keyword extraction.',
 'Document analysis involves extracting keywords.']

In [71]:
query

'Natural language processing techniques enhance keyword extraction efficiency.'

In [75]:
response = co.rerank(
    model="rerank-english-v3.0",
    query="Natural language processing techniques enhance keyword extraction efficiency.",
    documents=top_4_documents,
    return_documents=True
)

In [76]:
print(response)

id='b2fe59ce-dc2a-49a1-96b0-7d2cac38c5b1' results=[RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Efficient keyword extraction enhances search accuracy.'), index=0, relevance_score=0.99411184), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Machine learning algorithms can optimize keyword extraction methods.'), index=1, relevance_score=0.9129032), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Understanding document structure aids in keyword extraction.'), index=2, relevance_score=0.32885265), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Document analysis involves extracting keywords.'), index=3, relevance_score=0.02865267)] meta=ApiMeta(api_version=ApiMetaApiVersion(version='1', is_deprecated=None, is_experimental=None), billed_units=ApiMetaBilledUnits(images=None, input_tokens=None, image_tokens=None, output_tokens=None, search_units=1.0, classifications=None), tokens=Non

In [77]:
from pprint import pprint

pprint(response)

RerankResponse(id='b2fe59ce-dc2a-49a1-96b0-7d2cac38c5b1', results=[RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Efficient keyword extraction enhances search accuracy.'), index=0, relevance_score=0.99411184), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Machine learning algorithms can optimize keyword extraction methods.'), index=1, relevance_score=0.9129032), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Understanding document structure aids in keyword extraction.'), index=2, relevance_score=0.32885265), RerankResponseResultsItem(document=RerankResponseResultsItemDocument(text='Document analysis involves extracting keywords.'), index=3, relevance_score=0.02865267)], meta=ApiMeta(api_version=ApiMetaApiVersion(version='1', is_deprecated=None, is_experimental=None), billed_units=ApiMetaBilledUnits(images=None, input_tokens=None, image_tokens=None, output_tokens=None, search_units=1.0, classifications=

In [78]:
response.results[0].document.text

'Efficient keyword extraction enhances search accuracy.'

In [79]:
response.results[0].relevance_score

0.99411184

In [80]:
for i in range(4):
  print(f'text: {response.results[i].document.text} score: {response.results[i].relevance_score}')

text: Efficient keyword extraction enhances search accuracy. score: 0.99411184
text: Machine learning algorithms can optimize keyword extraction methods. score: 0.9129032
text: Understanding document structure aids in keyword extraction. score: 0.32885265
text: Document analysis involves extracting keywords. score: 0.02865267


## Now comparing reranked results from BM25, Cross Encoder and Cohere

# Original top 4 documents using Sentence Transformer

Top 4 Documents:

Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 0.7521412372589111

Rank 2: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.7448166012763977

Rank 3: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - 0.6316117644309998

Rank 4: Document - 'Document analysis involves extracting keywords.', Similarity Score - 0.5675694346427917

# BM25 Reranking (Top 4 Documents)

Rerank of Top 4 Documents:

Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 0.19079979534096053

Rank 2: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - 0.1780325227902643

Rank 3: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.1668667199671815

Rank 4: Document - 'Document analysis involves extracting keywords.', Similarity Score - 0.0

# Cross-Encoder Reranking (Top 4 Documents)

Rerank of Top 4 Documents:

Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Similarity Score - 3.1378731727600098

Rank 2: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Similarity Score - 0.8421645760536194

Rank 3: Document - 'Document analysis involves extracting keywords.', Similarity Score - -2.8781914710998535

Rank 4: Document - 'Understanding document structure aids in keyword extraction.', Similarity Score - -2.9193015098571777

# Cohere API Reranking (Top 4 Documents)

Rerank of Top 4 Documents:

Rank 1: Document - 'Efficient keyword extraction enhances search accuracy.', Relevance Score - 0.99411184

Rank 2: Document - 'Machine learning algorithms can optimize keyword extraction methods.', Relevance Score - 0.9129032

Rank 3: Document - 'Understanding document structure aids in keyword extraction.', Relevance Score - 0.32885265

Rank 4: Document - 'Document analysis involves extracting keywords.', Relevance Score - 0.02865267

## Conclusion

Among the three reranking techniques, **Cohere API Reranker** produced the best results for this query. It ranked the most relevant documents with the highest confidence by understanding the semantic meaning and context of the query.

The **Cross-Encoder** also performed well and is an excellent choice for local, high-quality reranking without relying on an external API.

**BM25** is the fastest approach and works well for keyword matching, but it cannot capture semantic relationships as effectively as Cross-Encoder or Cohere.

**Overall Choice:** **Cohere API Reranker** because it provides the highest-quality reranking, making it the preferred option for production-grade RAG applications where retrieval accuracy is critical.